fhir-r4/src/main/java/org/dhis2/fhir/adapter/fhir/transform/dhis/impl/aggregate/r4/R4FhirMeasureReportToDataValueSetTransformer.java

```diff
@@ -0,0 +1,212 @@
+        @Nonnull ScriptExecutionContext scriptExecutionContext,
+        @Nonnull ValueConverter valueConverter )
+    {
+        super( scriptExecutor, trackedEntityMetadataService, organizationUnitService,
+            trackedEntityService, fhirDhisAssignmentRepository, scriptExecutionContext,
+            valueConverter );
+    }
+
+    @Nonnull
+    @Override
+    public Set<FhirVersion> getFhirVersions()
+    {
+        return FhirVersion.R4_ONLY;
+    }
+
+    @Override
+    public FhirToDhisTransformOutcome<DataValueSet> transformInternal( @Nonnull final FhirClientResource fhirClientResource,
+        @Nonnull final FhirToDhisTransformerContext context, @Nonnull final IBaseResource input,
+        @Nonnull final RuleInfo<DataValueSetRule> ruleInfo, @Nonnull final Map<String, Object> scriptVariables )
+        throws TransformerException
+    {
+        final Map<String, Object> variables = new HashMap<>( scriptVariables );
+
+        final DataValueSet dataValueSet = getResource( fhirClientResource, context, ruleInfo, scriptVariables ).orElse( null );
+        if ( dataValueSet == null )
+        {
+            return null;
+        }
+
+        final MeasureReport mr = (MeasureReport) input;
+
+        final Optional<OrganizationUnit> organizationUnit;
+        if ( ruleInfo.getRule().getOrgUnitLookupScript() == null )
+        {
+            logger.info( "Rule does not define an organization unit lookup script and data value set does not yet include one." );
+            return null;
+        }
+        else
+        {
+            organizationUnit = getOrgUnit( context, ruleInfo, ruleInfo.getRule().getOrgUnitLookupScript(), variables );
+            organizationUnit.ifPresent( ou -> dataValueSet.setOrgUnitId( ou.getId() ) );
+        }
+
+        if ( !organizationUnit.isPresent() )
+        {
+            return null;
+        }
+
+        //------------------------------
+        if ( ruleInfo.getRule().getDataSetIdLookupScript() == null )
+        {
+            logger.info( "Rule does not define a data value set lookup script and data value set does not yet include one." );
+            return null;
+        }
+        else
+        {
+            Optional<String> dataSetId = getDataSetId( context, ruleInfo, ruleInfo.getRule().getDataSetIdLookupScript(), variables );
+            dataSetId.ifPresent( dataValueSet::setDataSetId );
+        }
+
+        if ( dataValueSet.getDataSetId() == null )
+        {
+            return null;
+        }
+        //------------------------------
+
+        ZonedDateTime lastUpdated = ZonedDateTime.ofInstant( mr.getMeta().getLastUpdated().toInstant(), zoneId );
+        //        String version = mr.getMeta().getVersionId();
+
+        dataValueSet.setId( createArtificialDataValueSetId( mr ) );
+        dataValueSet.setLastUpdated( lastUpdated );
+        dataValueSet.setPeriod( PeriodUtils.getDHIS2PeriodString( mr.getPeriod().getStart(), mr.getPeriod().getEnd() ) );
```

Comment:
Period start and end are optional in FHIR. OpenLMIS profile will not be validated by the Adapter (in General profile checking could be added, but this has other advantages/disadavantages).